pysb/simulator/ssa_base.py

```diff
@@ -0,0 +1,151 @@
+                              expr_strings[e.name],
+                              rate)
+
+            # replace x**2 with (x-1)*x
+            pattern = "(_{2}s\d+)\*\*(\d+)"
+            matches = re.findall(pattern, rate)
+            for m in matches:
+                repl = m[0]
+                for i in range(1, int(m[1])):
+                    repl += "*(%s-%d)" % (m[0], i)
+                rate = re.sub(pattern, repl, rate)
+            # replace species string with matrix index (`_si` with `y[i]`)
+            rate = re.sub(r'_{2}s(\d+)', lambda m: 'y[%s]' % (int(m.group(1))),
+                          rate)
+            # replace param names with vector notation
+            for q, prm in enumerate(params_names):
+                rate = re.sub(r'\b(%s)\b' % prm, 'param_vec[%s]' % q, rate)
+
+            # Calculate the fast approximate, better performance on GPUs
+            rate = rate.replace('pow', 'powf')
+            # If a parameter is a float and appears first, the result output
+            # will lose precision. Casting to double ensures precision
+            rate = '(double)' + rate
+
+            output_string += rate + ";\n"
+        return dict(n_species=self._n_species, n_params=self._parameter_number,
+                    n_reactions=_reaction_number, propensities=output_string,
+                    stoch=stoich_string)
+
+    def run(self, tspan=None, param_values=None, initials=None, number_sim=0):
+
+        num_sim = int(number_sim)
+
+        # check for proper arguments
+        if param_values is None and initials is None and not num_sim:
+            raise SimulatorException("Please provide a multi-dimension set of "
+                                     "parameters, initials, or number_sim>0")
+        elif param_values is None and not num_sim:
+            self.initials = initials
+            num_sim = self.initials.shape[0]
+        elif initials is None and not num_sim:
+            self.param_values = param_values
+            num_sim = self.param_values.shape[0]
+
+        if param_values is None and initials is None:
+            # Run simulation using same param_values
+            # initials will be taken care of on SimulatorBase side
+            param_values = np.repeat(self.param_values, num_sim, axis=0)
+        elif isinstance(param_values, np.ndarray) and len(
+                param_values.shape) == 1:
+            # initials taken care of on SimulatorBase side
+            param_values = np.repeat([param_values], num_sim, axis=0)
+        elif isinstance(initials, np.ndarray) and len(initials.shape) == 1:
+            # parameters taken care of on SimulatorBase side
+            initials = np.repeat([initials], num_sim, axis=0)
+        self.num_sim = num_sim
+        super(SSABase, self).run(tspan=tspan, initials=initials,
+                                 param_values=param_values,
+                                 _run_kwargs=locals())
+
+    @staticmethod
+    def _get_stoich(model):
```

Comment:
Could you use `model.stoichiometry_matrix` instead? It's very new and it definitely wasn't available when you started this. I already noticed the code here doesn't handle fixed-concentration species so that's one feature you'd get for free by adopting it.

https://github.com/pysb/pysb/blob/master/pysb/core.py#L2109

Reply:
Nice! Updated